src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/SemanticTokensCache.cs

```diff
@@ -0,0 +1,210 @@
+        /// of the range.
+        /// </summary>
+        /// <param name="uri">The URI associated with the cached tokens.</param>
+        /// <param name="semanticVersion">The semantic version associated with the cached tokens.</param>
+        /// <param name="requestedRange">The requested range for the desired tokens.</param>
+        /// <param name="cachedTokens">If found, contains the cached range and cached tokens.</param>
+        /// <returns></returns>
+        internal bool TryGetCachedTokens(
+            DocumentUri uri,
+            VersionStamp semanticVersion,
+            Range requestedRange,
+            [NotNullWhen(true)] out (Range Range, List<int> Tokens)? cachedTokens)
+        {
+            if (!_cache.TryGetValue(uri, out var documentCache) ||
+                !documentCache.TryGetValue(semanticVersion, out var lineToTokensDict))
+            {
+                // No cached results found
+                cachedTokens = null;
+                return false;
+            }
+
+            var tokens = new List<int>();
+
+            // Keep track of where the cached range begins
+            Position? cachedRangeStart = null;
+
+            // Keep track of where the cached range ends
+            var cachedRangeNumLines = 0;
+
+            // Keeps track of the current empty line count (needed for calculating line offsets)
+            var emptyLineCount = 0;
+
+            // Retrieve the cached tokens associated with the passed-in range
+            for (var line = requestedRange.Start.Line; line < requestedRange.End.Line; line++)
+            {
+                if (!lineToTokensDict.TryGetValue(line, out var lineTokens))
+                {
+                    if (tokens.Count != 0)
+                    {
+                        // We already have some sort of cached results, but we've now reached
+                        // a gap in the cache which we have no tokens for.
+                        // Since we only want to return a continuous range, we'll exit early.
+                        break;
+                    }
+
+                    continue;
+                }
+
+                if (cachedRangeStart is null)
+                {
+                    cachedRangeStart = new Position { Line = line, Character = 0 };
+                }
+
+                cachedRangeNumLines++;
+
+                // If a line's List is empty, it means the line has no tokens
+                if (lineTokens.Count == 0)
+                {
+                    emptyLineCount++;
+                    continue;
+                }
+
+                // We've found a non-empty cached line. Go through and process it.
+                for (var lineTokenIndex = 0; lineTokenIndex < lineTokens.Count; lineTokenIndex++)
+                {
+                    // If we're at the start of the cached range, the line offset should be relative
```

Comment:
Is this strictly true? What if we're at the start of the cached range but that's NOT the start of the tokens we've already cached for the document?

Reply:
I believe it's true for all cases. What's important here is the result we're returning to the caller, since they don't know which tokens we have/have not cached for the document. They just know about the cached range and tokens we return to them.